### DISCUSSION TITLE
Title: Using trial.report() with max value instead of current value Before trial.should_prune() for Correct Value Reporting in Pruned Trials

### DISCUSSION THREAD

**Discussion by ggenesum** [optuna/optuna#4952] [Q&A]
### Don't use GitHub Issues to ask support questions.

Hello Optuna team,

I've been using Optuna for hyperparameter optimization and have run into an interesting situation regarding pruned trials.

Typically, when a trial should be pruned, we'd call:

```
trial.report(accuracy_at_epoch_i, step=i)

if trial.should_prune():
    raise optuna.exceptions.TrialPruned()

```
However, in my scenario, I'm tracking the best value across epochs, and I'd like to ensure that this best value is reported back to the trial before it gets pruned. So I was thinking of the following approach:

```
# ... training loop ...
best_accuracy_accross_epochs = max(accuracy_list)
# Report the current best value
trial.report(best_accuracy_accross_epochs, step=i)

# Check if the trial should be pruned
if trial.should_prune():
    raise optuna.exceptions.TrialPruned()

# ... rest of the training loop ...

```

With this, I hope to keep Optuna informed about the best value for the trial up to the point of pruning, and then also indicate that the trial was pruned. My main goal is to ensure that the best observed value is correctly stored with the pruned trial in the Optuna study database.

Is this approach recommended and valid? Will this impact the pruners or samplers in any unexpected ways?

Thanks for your help and for creating such a valuable tool!

**Comment by nzw0301** [comment]
I think reporting the best intermediate value just before pruning is not recommended. The reported value is used when we call `trial.should_prune`, so that will cause unexpected comparison for pruning.
